src/test/spec/unified_runner/operation.rs

```diff
@@ -136,21 +137,47 @@ macro_rules! with_mut_session {
     ($test_runner:ident, $id:expr, |$session:ident| $body:expr) => {
         async {
             let id = $id;
-            let mut session_owned = match $test_runner.entities.write().await.remove(id).unwrap() {
-                Entity::Session(session_owned) => session_owned,
+            let entity = $test_runner.entities.write().await.remove(id).unwrap();
+            match entity {
+                Entity::Session(mut session_owned) => {
+                    let $session: &mut ClientSession = &mut session_owned;
+                    let out = $body.await;
+                    $test_runner
+                        .entities
+                        .write()
+                        .await
+                        .insert(id.to_string(), Entity::Session(session_owned));
+                    out
+                }
+                Entity::SessionPtr(ptr) => {
+                    let $session = unsafe { &mut *ptr.0 };
+                    let out = $body.await;
+                    $test_runner
+                        .entities
+                        .write()
+                        .await
+                        .insert(id.to_string(), Entity::SessionPtr(ptr));
+                    out
+                }
                 o => panic!(
                     "expected {} to be a session entity, instead was {:?}",
                     $id, o
                 ),
-            };
-            let $session = &mut session_owned;
-            let out = $body.await;
-            $test_runner
-                .entities
-                .write()
-                .await
-                .insert(id.to_string(), Entity::Session(session_owned));
-            out
+            }
+        }
+    };
+}
+
+macro_rules! with_opt_session {
```

Comment:
This is an unrelated convenience; many of the operation impls work in exactly the same way, so this captures that pattern and makes 'em easier to read.